internal/plugin/sdk/host/interface.go

```diff
@@ -0,0 +1,83 @@
+// Package host provides the interfaces and functionality for host
+// plugins. Host plugins provide functionality for dynamic host
+// catalogs and host sets.
+package host
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/plugin/sdk/proto"
+)
+
+// HostPlugin describes the interface for host plugins.
+type HostPlugin interface {
+	// OnCreateCatalog is a hook that runs when a host catalog is
+	// created. It takes a context and the host catalog to create,
+	// which may contain (optional) secret data to help authenticate
+	// the request against a cloud API.
+	//
+	// The plugin may return optional state data to be persisted and
+	// at-rest encrypted. This should be used to store authentication
+	// data and other necessary configuration to be used in later hooks
+	// and lookup calls. If an error is returned, data persistence is
+	// aborted.
+	OnCreateCatalog(ctx context.Context, cat proto.HostCatalog) (proto.HostCatalogPersisted, error)
+
+	// OnUpdateCatalog is a hook that runs when a host catalog is
+	// updated. It takes a context and the host catalog to update, both
+	// in its pre-update, and post-update form. The post-update catalog
+	// may contain (optional) secret data. The existing persisted state
```

Comment:
I think we need to work on figuring out our approach for updating secrets.  Since this is a write only field in the API do we 1. require all secret values to be provided every time the secrets field is updated or 2. do we allow partial updates of the secrets field?

The benefit of approach 1 is that it sort of matches what we are doing with the host catalog in that it provides the complete old and the complete new, assuming the plugin saves the complete old in the persisted data.

I like approach 2 because it matches our existing APIs approach for updating fields (although we've never really had a write only object w/ updatable subfields before).  Following this route we'll need to also pass in an update mask for secrets in the update catalog calls to the plugin.

@jefferai you and I have discussed update semantics w/ the API quite a bit.  Do you have an opinion here?

Reply:
I also think there's something to be said about the simplicity of the first approach, but yeah, let's see what Jeff has to say.

Reply:
I like approach 1. I think that what's in secrets should be extremely minimal (e.g. secret key only, not key ID which would live in attributes) so the only time you'd actually write it is to explicitly provide a credential, and it's not worth dealing with update masks for that case.